superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx

```diff
@@ -122,24 +126,62 @@ const EDIT_DATASET = 'edit_dataset';
 const QUERY_PREVIEW = 'query_preview';
 const SAVE_AS_DATASET = 'save_as_dataset';
 
+// If the string is longer than this value's number characters we add
+// a tooltip for user can see the full name by hovering over the visually truncated string in UI
+const VISIBLE_TITLE_LENGTH = 25;
+
+// Assign icon for each DatasourceType.  If no icon assingment is found in the lookup, no icon will render
+export const datasourceIconLookup = {
+  [DatasourceType.Query]: (
+    <Icons.ConsoleSqlOutlined className="datasource-svg" />
+  ),
+  [DatasourceType.Table]: <Icons.DatasetPhysical className="datasource-svg" />,
+};
+
+// Render title for datasource with tooltip only if text is longer than VISIBLE_TITLE_LENGTH
+export const renderDatasourceTitle = (displayString, tooltip) =>
+  displayString.length > VISIBLE_TITLE_LENGTH ? (
+    // Add a tooltip only for long names that will be visually truncated
+    <Tooltip title={tooltip}>
+      <span className="title-select">{displayString}</span>
+    </Tooltip>
+  ) : (
+    <span title={tooltip} className="title-select">
+      {displayString}
+    </span>
+  );
+
+// Different data source types use different attributes for the display title
+export const getDatasourceTitle = datasource =>
+  datasource?.sql ?? datasource?.name ?? '';
+
+// When this is a Query with SQL, we wnt the SQL to be the tooltip
+export const getTooltip = datasource =>
+  datasource?.sql ?? datasource?.name ?? '';
+
 class DatasourceControl extends React.PureComponent {
   constructor(props) {
     super(props);
     this.state = {
       showEditDatasourceModal: false,
       showChangeDatasourceModal: false,
+      showSaveDatasetModal: false,
     };
-    this.onDatasourceSave = this.onDatasourceSave.bind(this);
-    this.toggleChangeDatasourceModal =
-      this.toggleChangeDatasourceModal.bind(this);
-    this.toggleEditDatasourceModal = this.toggleEditDatasourceModal.bind(this);
-    this.toggleShowDatasource = this.toggleShowDatasource.bind(this);
-    this.handleMenuItemClick = this.handleMenuItemClick.bind(this);
-    this.toggleSaveDatasetModal = this.toggleSaveDatasetModal.bind(this);
   }
 
-  onDatasourceSave(datasource) {
-    this.props.actions.changeDatasource(datasource);
+  getDatasourceAsSaveableDataset = source => {
+    const dataset = {
+      columns: source?.columns || [],
+      name: source?.datasource_name || source?.name || t('Untitled'),
+      dbId: source?.database.id,
+      sql: source?.sql || '',
+      schema: source?.schema,
+    };
+    return dataset;
```

Comment:
nit, but is the const necessary? I think you can just return the object that you're defining.